### DISCUSSION TITLE
dev setup ?

### DISCUSSION THREAD

**Discussion by maxandersen** [moralmunky/Home-Assistant-Mail-And-Packages#734] [Q&A]
Hi,

I really like the idea of mail driven package tracking so want to add a bunch of the shippers I use here in Denmark - but I'm not sure which dev setup needed for it.

is it just about installing tox and run it ?

I see test.py that seem to require modifying username/passwords ?

Any tips on how to best test/development on this would be appreciated :)

**Comment by firstof9** [comment]
Tox is for testing.

The `test.py` is an old test, you'll likely not want to use that.

To dev, I just use VSCode for coding, testing for new shippers is pretty easy to setup in the tests, just use the current test as examples.

For verification, I just spin up a dev branch version of Home Assistant and drop the modified code into it to make sure everything's working as expected, but I believe you can spin up a container from VSCode as well to test it locally.

**Reply by maxandersen** [reply]
so just adjust tests and run using "tox test" ? ( I dont see way to run the tests from vscode )

**Reply by firstof9** [reply]
just `tox` but make sure you have `tox` installed, similar to the Home Assistant dev environment setup.

https://developers.home-assistant.io/docs/development_environment
